File: serve.mjs
// Local server for First Return. Serves docs/ with the cross-origin isolation headers that let the
// depth model use several CPU threads. Run: node serve.mjs   (add --no-browser to skip opening it)
import http from 'node:http';
import fs from 'node:fs';
import path from 'node:path';
import os from 'node:os';
import { exec } from 'node:child_process';
import { fileURLToPath } from 'node:url';

const PORT = 8811;
const ROOT = path.join(path.dirname(fileURLToPath(import.meta.url)), 'docs');
const TYPES = { '.html':'text/html; charset=utf-8', '.js':'text/javascript', '.mjs':'text/javascript', '.wasm':'application/wasm',
  '.json':'application/json', '.webmanifest':'application/manifest+json', '.svg':'image/svg+xml', '.png':'image/png', '.jpg':'image/jpeg',
  '.onnx':'application/octet-stream', '.tflite':'application/octet-stream', '.css':'text/css' };

http.createServer((req, res) => {
  let p = decodeURIComponent(new URL(req.url, 'http://x').pathname);
  if (p.endsWith('/')) p += 'index.html';
  const file = path.normalize(path.join(ROOT, p));
  if (!file.startsWith(ROOT)) { res.writeHead(403); return res.end(); }
  fs.stat(file, (err, st) => {
    if (err || !st.isFile()) { res.writeHead(404); return res.end('Not found'); }
    res.writeHead(200, {
      'Content-Type': TYPES[path.extname(file).toLowerCase()] || 'application/octet-stream',
      'Content-Length': st.size,
      'Cache-Control': 'no-cache',
      'Cross-Origin-Opener-Policy': 'same-origin',
      'Cross-Origin-Embedder-Policy': 'credentialless',
    });
    if (req.method === 'HEAD') return res.end();
    fs.createReadStream(file).pipe(res);
  });
}).listen(PORT, '0.0.0.0', () => {
  console.log(`First Return is running.\n  On this computer: http://localhost:${PORT}`);
  for (const list of Object.values(os.networkInterfaces())) for (const a of list || [])
    if (a.family === 'IPv4' && !a.internal) console.log(`  On your phone (same Wi-Fi): http://${a.address}:${PORT}`);
  console.log('Close this window to stop it.');
  if (!process.argv.includes('--no-browser')) exec(`start "" http://localhost:${PORT}`);
});

File: serve.py
# Local server for First Return. Cross-origin isolation headers let the depth model use several CPU threads.
import http.server, socket, sys, os, webbrowser, threading
PORT = 8811
os.chdir(os.path.join(os.path.dirname(os.path.abspath(__file__)), "docs"))
class H(http.server.SimpleHTTPRequestHandler):
    extensions_map = {**http.server.SimpleHTTPRequestHandler.extensions_map,
        ".mjs": "text/javascript", ".js": "text/javascript", ".wasm": "application/wasm",
        ".webmanifest": "application/manifest+json", ".svg": "image/svg+xml"}
    def end_headers(self):
        self.send_header("Cross-Origin-Opener-Policy", "same-origin")
        self.send_header("Cross-Origin-Embedder-Policy", "credentialless")
        self.send_header("Cache-Control", "no-cache")
        super().end_headers()
    def log_message(self, *a): pass
def lan_ip():
    try:
        s = socket.socket(socket.AF_INET, socket.SOCK_DGRAM); s.connect(("8.8.8.8", 80)); ip = s.getsockname()[0]; s.close(); return ip
    except Exception: return None
srv = http.server.ThreadingHTTPServer(("0.0.0.0", PORT), H)
print(f"First Return is running.\n  On this computer: http://localhost:{PORT}")
ip = lan_ip()
if ip: print(f"  On your phone (same Wi-Fi): http://{ip}:{PORT}")
print("Close this window to stop it.")
if "--no-browser" not in sys.argv: threading.Timer(0.8, lambda: webbrowser.open(f"http://localhost:{PORT}")).start()
srv.serve_forever()
